Rumble video embed

benalleng opened the pull request:
Closes #134 but if this works as we hope I am hoping to add in the embed functionality for other sites as well

Embeds rumble video embed links, unfortunately rumble is weird and what you see at the top is not the embed link you want like it is on youtube but thats the price you pay i guess for their revenue share model. 
It has to do with the fact that embeded videos have a different revenue share percentage so keeping track of how to divide the money is too difficult when everyone is downloading the video from the same url (I'm guessing)

Rumble vs a Youtube embed
![image](https://github.com/stackernews/stacker.news/assets/108441023/a3af9625-4b0e-479c-8779-4d1a7395245a)

This standard `<iframe/>` model should be able to be used on many sites so maybe once we do some quick QA on this I could try and add a more general "video embed" that parses video embeds from several of the other popular video sites


I'm going to have this PR open for now but if we want to add embedding for more sites i can put it into draft untl the logic is finished for those as well

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
	- Introduced a new component for enhanced link rendering, including special handling for YouTube and Rumble video links.
- **Security Updates**
	- Updated content security policy to allow embedding from `rumble.com`, alongside existing permissions for `www.youtube.com` and `platform.twitter.com`.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

benalleng commented on components/text.js:
```diff
@@ -163,66 +237,7 @@ export default memo(function Text ({ rel, imgproxyUrls, children, tab, itemId, o
-            // assume the link is an image which will fallback to link if it's not
-            return <Img src={href} rel={rel ?? UNKNOWN_LINK_REL} {...props}>{children}</Img>
-          },
+          a: A,
```
The logic here was getting quite cumbersome so I decided to move it up to where all the other elements were defined

ekzyis commented on components/text.js:
```diff
@@ -137,6 +137,80 @@ export default memo(function Text ({ rel, imgproxyUrls, children, tab, itemId, o
+    }
+
+    // if the link is to a rumble video, render the video
+    const rumble = href.match(/(?:https?:\/\/)?(?:www\.)?rumble\.com\/embed\/(?<id>[a-z0-9]+)\/\?pub=\d+&start=(?<start>\d+)/i)
```
I tested with this URL: https://rumble.com/embed/v17j3tt/?pub=3fid3f

As you can see, `pub` can be alphanumeric so you want to use `[a-z0-9]+` or `\w+` for it. And  `start` should not be required to match.

I noticed that we could probably leverage the URL constructor here for parsing [as I did in `parseInternalLinks`](https://github.com/stackernews/stacker.news/blob/b4cef44a43bcdb802e1d035748c84a420aa22218/lib/url.js#L29-L52) (see https://github.com/stackernews/stacker.news/pull/765#issuecomment-1907650129). I think this would make this easier and cleaner.

The other twitter and youtube embed code you can see here is just old and bad and isn't using them yet.

ekzyis reviewed (changes requested):
Thanks for the PR!

That it doesn't work for direct links is unfortunate. Did you check if there is a way to fetch the embed link from direct links? If not possible, that's okay. Maybe that's even too complicated so not worth it. I think stackers will learn fast to use embed links with rumble.

I noticed three other things though:

1. The regexp is not matching for links without `start` and for links where `pub` contains letters.
2. We also want this to work for link posts. This currently only works for links inside a post description or comment. So the rumble embed code also needs to be added to `ItemEmbed` which you can find in components/item-full.js. That means (more) duplicate code though so if you want, you can try to refactor it. Not necessary for this PR though, I usually like create separate PRs (["stacked PRs"](https://benjamincongdon.me/blog/2022/07/17/In-Praise-of-Stacked-PRs/)) for stuff like this anyway.

![2024-03-27-153441_1920x1080_scrot](https://github.com/stackernews/stacker.news/assets/27162016/0d33eba3-9b59-49d9-938d-2d0684c247e5)

7. I think we don't need to use error-prone regexp's but can use the URL constructor. See my comment.

benalleng commented:
Thanks for the feedback! Regex creation is  definitely a skill I have not grasped yet lol I'll take a look at the refactor as well if we intend to add embeds for many different sites it might be worth pulling them up to be reused in multiple places

> Did you check if there is a way to fetch the embed link from direct links?

Short of scraping the data somehow when loading in from the standard link I don't think its possible. in this video https://rumble.com/v1a59rb-rumble-basics-how-to-embed-your-video.html they talk about how embeded videos get an 80% revenue share so I imagine they don't want us to be able to claw back that 20% on a simple script

> If you want to refactor the existing embed code to also use the URL constructor for parsing, I recommend doing this in a separate PR.

Ok I will try to get this rumble embed right first and then I can come back through with a refactor

huumn commented:
Nice iframe solution!

> Embeds rumble video embed links, unfortunately rumble is weird and what you see at the top is not the embed link you want like it is on youtube but thats the price you pay i guess for their revenue share model.

I don't think we can expect people to share the right url, so if we want this to be used, we'll probably want to make it work for either url:

1. if it's the embed url, use your solution as is
2. if it's the url someone is likely to copy and paste onto SN, fetch the url on the client and parse out the embed link, then replace the copied url

benalleng commented:
>     2. if it's the url someone is likely to copy and paste onto SN, fetch the url on the client and parse out the embed link, then replace the copied url

Ok! /yeah i think i was overthinking it, good idea i'll give that a try!

benalleng commented:
Converting to draft for now until I can get a better sense of how to fetch the embed link.

ekzyis commented:
>> fetch the url on the client and parse out the embed link, then replace the copied url
>
> Ok! yeah i think i was overthinking it, good idea! i'll give that a try!

I am confused. Maybe my English was confusing but that's what I meant with

>> Did you check if there is a way to fetch the embed link from direct links?
>
> Short of scraping the data somehow when loading in from the standard link I don't think its possible. [...]

However, I forgot that we're already scraping websites to auto-populate titles + publication date [here](https://github.com/stackernews/stacker.news/blob/4c2fcec69be0432cf3641a49397e448ed0a356d0/api/resolvers/item.js#L533-L554) with [`domino`](https://www.npmjs.com/package/domino), a server-side DOM implementation. Might be interesting code to look at for you. (It's on the server though.)

benalleng commented:
> I am confused. Maybe my English was confusing but that's what I meant with

No you're fine! That's on me I just wasn't thinking about a tool like domino

benalleng commented on components/item-full.js:
```diff
@@ -98,6 +112,47 @@ function ItemEmbed ({ item }) {
+        </div>
+      )
+    } else {
+      return null
```
added these return nulls for now As i was hoping to create some logic for when the user's link does Not follow the expected embed format but at the moment no logic for that in rumble

benalleng commented:
I cleaned up the logc for rumble and added the embeds for link posts as well.  Also in my attempts to find a solution for our embed vs non-embeded links I realized odysee handles embeds more simply.

I am a little stuck with the client-side fetching, I either am unable to get the tool to work like domino or puppeteer as they seem to be designed for server-side or when trying to fetch the data I am getting CORS errors when trying to grab the data directly when clientside. @ekzyis any thoughts on another avenue I could try? I breifly looked at the naming convention between `https://rumble.com/embed/v17j3tt/?pub=3fid3f` and `https://rumble.com/v1a59rb-rumble-basics-how-to-embed-your-video.html` but I couldn't find any way to decode or parse one id from the other.

huumn commented:
I'd think you can just `fetch` the page then regex parse for `/embed/` or something in the body.

benalleng commented on api/resolvers/item.js:
```diff
@@ -553,6 +553,17 @@ export default {
+        const response = await fetch(ensureProtocol(url))
+        const html = await response.text()
+        if (url.includes('peertube.tv')) {
+          return html.match(/"([^"]*\/embed\/[^"]*)"/)[0].substring(1, html.match(/"([^"]*\/embed\/[^"]*)"/)[0].length - 1)
```
This is the parse for the peertube html, it looks for the embed link and then trims the `" "` on either end

benalleng replied:
From my brief investigation becauses `bitcoin.tv` is based on peertube it should have the exact same format so we should be able to use this exact code for parsing links from there as well

benalleng commented on api/resolvers/item.js:
```diff
@@ -553,6 +553,17 @@ export default {
     },
+    fetchDocument: async (parent, { url }, { models }) => {
+      try {
+        const response = await fetch(ensureProtocol(url))
```
The really confusing thing is that when I test a get request with this rumble link in something like postman, it works

benalleng replied:
Here is the error object being thrown
```
{
	"errors": [
		{
			"message": "fetch failed",
			"locations": [
				{
					"line": 2,
					"column": 3
				}
			],
			"path": [
				"fetchDocument"
			],
			"extensions": {
				"code": "INTERNAL_SERVER_ERROR",
				"stacktrace": [
					"Error: fetch failed",
					"    at Object.fetchDocument (webpack-internal:///(api)/./api/resolvers/item.js:578:23)",
					"    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
				]
			}
		}
	],
	"data": {
		"fetchDocument": null
	}
}
```

huumn replied:
Wow that's not descriptive at all. Regardless, I think we want to call this on the client.

Have you tried `fetch`ing in your browser console?

benalleng replied:
When I fetch from the frontend or the console I get CORS errors
`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://rumble.com/v1a59rb-rumble-basics-how-to-embed-your-video.html. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.`
`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://rumble.com/v1a59rb-rumble-basics-how-to-embed-your-video.html. (Reason: CORS request did not succeed). Status code: (null).
`

huumn replied:
Ah, that makes sense. CORs is a pain.

I'd try getting a more descriptive error on the server.

benalleng replied:
I was having a look at #879 and I was curious if rumble had a similar cors policy so I gave it a test by using this https://github.com/Rob--W/cors-anywhere as a test to see if I could grab the embed link and it works!

Note the `html` link but it still renders the embeded video
![image](https://github.com/stackernews/stacker.news/assets/108441023/cfd41c6e-d3c3-4ff6-96d4-b0c3413a1bbe)

benalleng replied:
didn't implement it in the sn repo yet, I just added the demo https://cors-anywhere.herokuapp.com/ server as a prefix to the url

huumn commented:
Closing after #1191 was merged
